### PR TITLE
Allow when condition on field validator

### PR DESCRIPTION
**Pull request by dimagorelik**
Allow when condition on field validator.
The change will allow to check condition on Parent entities before running validation

### CODE REVIEW

**Review by larisa-bud — APPROVED** [review]

File: main/src/main/java/com/kenshoo/pl/entity/spi/FieldValidator.java [hunk 1]
```diff
@@ -25,4 +29,18 @@
      * @return validation error or <code>null</code> if none
      */
     ValidationError validate(T fieldValue);
+
+    /**
+     * @return a list of fields to fetch. May contain fields of parent entities only
+     */
+    default Stream<EntityField<?, ?>> fetchFields() {
```
Comment (effiban) [inline main/src/main/java/com/kenshoo/pl/entity/spi/FieldValidator.java:36]: You probably copied this from `ImmutableFieldValidator` - but I think a better name would be `parentFieldsToFetch()`.
I remember discussing this name already some time ago.....

File: integration-tests/src/test/java/com/kenshoo/pl/entity/PersistenceLayerTest.java [hunk 2]
```diff
@@ -1234,6 +1252,19 @@ public ValidationError validate(Integer fieldValue) {
         }
     }
 
+    private static class TestField2WihParentConditionValidator extends TestField2Validator {
```
Comment (effiban) [inline integration-tests/src/test/java/com/kenshoo/pl/entity/PersistenceLayerTest.java:1255]: typo: WthParent -> W**i**thParent

File: integration-tests/src/test/java/com/kenshoo/pl/entity/PersistenceLayerTest.java [hunk 3]
```diff
@@ -573,6 +575,22 @@ public void validField2InvalidValueForCreateValidation() {
...
+
+        EntityChangeCompositeValidator<EntityForTest> changesCompositeValidator = new EntityChangeCompositeValidator<>();
+        changesCompositeValidator.register(new TestField2WihParentConditionValidator());
+        ChangeFlowConfig<EntityForTest> flowConfig = changeFlowConfig().withValidator(changesCompositeValidator).build();
+
+        UpdateResult<EntityForTest, EntityForTest.Key> updateResult = persistenceLayer.update(commands, flowConfig);
+
+        Collection<ValidationError> validationErrors = updateResult.getErrors(commands.get(0));
+        assertThat(validationErrors.iterator().next().getErrorCode(), is(FIELD2_ERROR));
+    }
```
Comment (effiban) [inline integration-tests/src/test/java/com/kenshoo/pl/entity/PersistenceLayerTest.java:592]: Please also add a test for the case where the condition is not met (where the parent id doesn't match so the validator is not applied). 
I see you did that in a unit test, but would be nice to have this here as well if it doesn't take too long.

File: main/src/main/java/com/kenshoo/pl/entity/spi/FieldValidator.java (hunk 1 above)
Comment (dimagorelik) [reply main/src/main/java/com/kenshoo/pl/entity/spi/FieldValidator.java:36]: @effiban 
I didn't copy from ImmutableFieldValidator.
In all our API fetchFields means parent's fields.
We should fix all APIs id different initiative

Comment (effiban) [reply main/src/main/java/com/kenshoo/pl/entity/spi/FieldValidator.java:36]: ah ok, I forgot that

**Review by effiban — APPROVED** [review]

**Pull request merged**